src/org/scada_lts/mango/service/DataPointService.java

```diff
@@ -161,6 +163,43 @@ public DataPointVO getDataPoint(int id) {
 		return dpList;
 	}
 
+	/**
+	 * Get PointList with permission validation
+	 *
+	 * Get All DataPoints with specific access permission.
+	 * To validate that you need to provide a user that should be
+	 * checked and the access type.
+	 *
+	 * @param user - User to validate
+	 * @param type - Permissions.DataPointAccessTypes entry [Read = 1, Set = 2]
+	 * @return - Filtered list of DataPoints object
+	 */
+	public List<DataPointVO> getDataPointsWithPermissions(User user, int type) {
+		List<DataPointVO> dpList = dataPointDAO.getDataPoints();
+		if(!user.isAdmin()) {
+			return dpList.stream()
+					.filter(dp -> filterDpWithReadAccess(dp, user, type))
+					.collect(Collectors.toList());
+		} else {
+			return dpList;
+		}
+	}
+
+	private boolean filterDpWithReadAccess(DataPointVO dp, User user, int type) {
+		try {
+			if(type == Permissions.DataPointAccessTypes.READ) {
+				Permissions.ensureDataPointReadPermission(user, dp);
+			} else if (type == Permissions.DataPointAccessTypes.SET) {
+				Permissions.ensureDataPointSetPermission(user, dp);
+			} else {
+				return false;
+			}
+			return true;
+		} catch (PermissionException e) {
+			return false;
+		}
+	}
```

Comment:
Hey, I would suggest using the interfaces in the package org.scada_lts.permissions.service :
```
package org.scada_lts.permissions.service;

import com.serotonin.mango.vo.DataPointVO;
import com.serotonin.mango.vo.User;
import org.scada_lts.dao.DataPointDAO;
import org.scada_lts.dao.model.ScadaObjectIdentifier;

import java.util.List;

public class GetDataPointsWithAccess implements GetObjectsWithAccess<DataPointVO, User> {

    private final DataPointDAO dataPointDAO;

    public GetDataPointsWithAccess(DataPointDAO dataPointDAO) {
        this.dataPointDAO = dataPointDAO;
    }

    @Override
    public List<DataPointVO> getObjectsWithAccess(User object) {
        return dataPointDAO.selectDataPointsWithAccess(object.getId(), object.getUserProfile());
    }

    @Override
    public List<ScadaObjectIdentifier> getObjectIdentifiersWithAccess(User object) {
        return dataPointDAO.selectDataPointIdentifiersWithAccess(object.getId(), object.getUserProfile());
    }
}
```
In DataPointDAO methods:
selectDataPointsWithAccess
selectDataPointIdentifiersWithAccess

You would need to add the query to include profiles:

```
	private static final String COLUMN_NAME_USER_PROFILE_ID = "userProfileId";
	public static final String DATA_POINT_FILTERED_BASE_ON_USER_ID_USERS_PROFILE_ID_ORDER_BY_NAME = ""
			+ COLUMN_NAME_USER_ID+"=? or "
			+ "id in (select dpu."+COLUMN_NAME_DATA_POINT_ID+" from dataPointUsers dpu where dpu."+COLUMN_NAME_USER_ID+"=? and dpu."+COLUMN_NAME_PERMISSION+">?) or "
			+ "id in (select dpup."+COLUMN_NAME_DATA_POINT_ID+" from dataPointUsersProfiles dpup where dpup."+ COLUMN_NAME_USER_PROFILE_ID +"=? and dpup."+COLUMN_NAME_PERMISSION+">?) "
			+ "order by dp." + COLUMN_NAME_DATAPOINT_NAME;

```
and methods:

```
        public List<DataPointVO> selectDataPointsWithAccess(final int userId, int profileId) {
		return filtered(DATA_POINT_FILTERED_BASE_ON_USER_ID_USERS_PROFILE_ID_ORDER_BY_NAME, new Object[] { userId, userId, ShareUser.ACCESS_NONE, profileId ,ShareUser.ACCESS_NONE}, 0);
	}

	public List<ScadaObjectIdentifier> selectDataPointIdentifiersWithAccess(int userId, int profileId) {
		return DAO.getInstance().getJdbcTemp().query(DATA_POINT_IDENTIFIER_SELECT + " where " + DATA_POINT_FILTERED_BASE_ON_USER_ID_USERS_PROFILE_ID_ORDER_BY_NAME,
				new Object[] { userId, userId, ShareUser.ACCESS_NONE, profileId ,ShareUser.ACCESS_NONE},
				new ScadaObjectIdentifierRowMapper.Builder()
						.idColumnName(COLUMN_NAME_ID)
						.xidColumnName(COLUMN_NAME_XID)
						.nameColumnName(COLUMN_NAME_DATAPOINT_NAME)
						.build());
	}

```
then:
```
public List<DataPointVO> getDataPointsWithPermissions(User user) {
                if(user.isAdmin())
                        return dataPointDAO.getDataPoints();
		return new GetDataPointsWithAccess(dataPointDAO).getObjectsWithAccess(user);
	}
or
public List<ScadaObjectIdentifier> getDataPointsWithPermissions(User user) {
                if(user.isAdmin())
                        return dataPointDAO.getSimpleList();
		return new GetDataPointsWithAccess(dataPointDAO).getObjectIdentifiersWithAccess(user);
	}
```
In the future, we will add DataPointDaoWithCache implementations.

Perhaps the presented code will need to be corrected, it was not compiled.